src/net/ossl.cc

```diff
@@ -703,6 +709,38 @@ class session : public enable_shared_from_this<session>, public session_impl {
         return make_exception_future<>(ossl_error("fatal error during ssl shutdown"));
     }
 
+    void verify() {
+        // A success return code (0) does not signify if a cert was presented or not, that
+        // must be explicitly queried via SSL_get_peer_certificate
+        auto res = SSL_get_verify_result(_ssl.get());
+        if (was_certificate_presented() && res != X509_V_OK) {
+            sstring stat_str(X509_verify_cert_error_string(res));
+            auto dn = extract_dn_information();
+            if (dn) {
+                std::stringstream ss;
+                ss << stat_str;
+                if (stat_str.back() != ' ') {
+                    ss << ' ';
+                }
+                ss << "(Issuer=[" << dn->issuer << "], Subject=[" << dn->subject << "])";
+                stat_str = ss.str();
+            }
+            throw verification_error(stat_str);
+        }
+        if (!was_certificate_presented() && _type == session_type::SERVER) {
+            if (_creds->get_client_auth() == client_auth::REQUIRE) {
+                throw verification_error("no certificate presented");
+            }
+            return;
+        }
```

Comment:
The logic of `was_certificate_presented()` is kind of confusing, because it would return false even if a certificate _was_ presented and `SSL_get_verify_result()` returned `!X509_V_OK`

Maybe we can simply this a little bit:

```suggestion
        if (res != X509_V_OK) {
            sstring stat_str(X509_verify_cert_error_string(res));
            auto dn = extract_dn_information();
            if (dn) {
                std::stringstream ss;
                ss << stat_str;
                if (stat_str.back() != ' ') {
                    ss << ' ';
                }
                ss << "(Issuer=[" << dn->issuer << "], Subject=[" << dn->subject << "])";
                stat_str = ss.str();
            }
            throw verification_error(stat_str);
        } else if (SSL_get0_peer_certificate(_ssl.get() == nullptr) {
            if (_creds->get_client_auth() == client_auth::REQUIRE) {
                throw verification_error("no certificate presented");
            }
            return;
        }
```

Reply:
Also do you want to return at this point or go on to the next line for dn_callback?

Reply:
The return is analogous to what `tls.cc::verify` does